L58-python-async-api.md

```diff
@@ -0,0 +1,1741 @@
+
+    async def SayHello(self,
+                       request: helloworld_pb2_grpc.HelloRequest,
+                       context: grpc.aio.ServicerContext
+                ) -> helloworld_pb2_grpc.HelloReply:
+        await asyncio.sleep(1)
+        return helloworld_pb2.HelloReply(message="Hello, %s!" % request.name)
+
+server = grpc.aio.server()
+server.add_insecure_port(":50051")
+helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+async with grpc.aio.insecure_channel("localhost:50051") as channel:
+    stub = echo_pb2_grpc.EchoStub(channel)
+    response = await stub.Hi(echo_pb2.EchoRequest(message="ping"))
+
+    async for response in stub.StreamingHi(...):
+        process(response)
+```
+
+### New Streaming API
+
+Existing streaming API has usability problem that its logic complexes user
+application. For client side, it requires the request iterator to be defined
+above the line of call, and the consumption of responses to be coded below the
+line of invocation. So when the application wants to send request based on
+received response, it would be quite challenging and usually involves
+multi-threading, which is not optimal in Python world.
+
+For server side, the same problem remains for non-trivial streaming servicer
+handlers. The `yield` syntax constraint the message sending should happen in the
+handler itself, otherwise the user needs to code a dedicated iterator class. For
+larger application, it would be a pain to create such iterator class for each
+streaming handler.
+
+euroelessar@ points out several great reasons for the new streaming API:
+* Enable `with` statement for streaming API, which allows clean resource management;
+* Easier to buffer outbound messages when peer pushes back (e.g. `yield` will
+  stop the execution of current coroutine, but `call.write` can be scheduled in
+  event loop);
+* Enable application to have both pending read and pending write in one
+  coroutine, and react on the first finisher (with iterator syntax this will be
+  quite hard to get it right).
+
+So, this gRFC introduces a new pattern of streaming API that reads/writes
+message to peer with explicit call.
+
+#### Snippet For New Streaming API
+
+```Python
+### Client side
+with stub.StreamingHi() as streaming_call:
+  request = echo_pb2.EchoRequest(message="ping")
+  await streaming_call.write(request)
+  response = await streaming_call.read()
+  while response:  # or response is not grpc.aio.EOF
+      process(response)
+      response = await streaming_call.read()
+
+### Server side
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def StreamingHi(self,
+                          unused_request_iterator,
+                          context: grpc.aio.ServicerContext
+            ) -> None:
+        bootstrap_request = await context.read()
+        initialize_environment(bootstrap_request)
```

Comment:
qq: what is the initialize_environment() good for?

Reply:
`initialize_environment` is a pseudo code for application logic, it is not part of the API surface.

Reply:
I just think that that initialize_environment() and has_response() are very confusing - can you make the example clearer to make it obvious what they symbolize?  (or at least add a comment what the operations mean?)

Reply:
I replaced them with `...` to avoid future confusion.

Reply:
It's still not crystal clear IMHO, but it's an improvement.

Looking at the two different APIs, I'm not sure if there's actually any significant difference
in the way they read:
- first example uses context.read() and doesn't use the `unused_request_iterator` at all
- the second example uses the async request iterator

instead of context.read(),   I could just use `anext(request_iterator)` to read the messages one by one and the result will be exactly the same, right? If so, does it make sense to expose the context.read() method at all?

I agree that for writing (on the server side), just relying on async generator (`yield`) would be too clumsy for many applications, so providing 2 ways of writing makes sense. QQ: how does the python server knows if that you've chosen API1(context.write() or API2 (yield) for given method?

Reply:
btw, IIRC the "bootstrap request" in example 1 is there to demonstrate that the first API is there to support more complex scenarios (like the first request in the stream being of a different kind ~ perhaps in the 'bootstrap' sense), but it's not obvious from the example that this is the point. perhaps the example can contain a comment with that point `# Reader/Writer API allows to handle more complex semantics, such as handling the first request as "bootstrap" request and then processing other requests according to the configuration`. Btw that particular scenario is relatively easy to handle with `yield` as well so perhaps this might not be the best example to use.

```
# async generator with bootstrap request (pretty easy to implement)
def async_generator:
    async for req in request_iterator:
         if is_first_request:
             bootstrap(req)
            is_first_request = False
         else:
            process_request()
```

I think the `yield` model might be difficult to use when there's not a 1 to 1 correspondence between requests and responses or when there's more logic involved (interaction with other internal components or calling into other services) - just mentioning that if we want to demonstrate what each of the APIs is useful for, we should use examples that really reflect that.

Reply:
+@euroelessar who is the advocator of reader / writer API.

Sorry I didn't response timely, I think GitHub is not ideal for long threads like this. Critique gets one point for complex reviews.

> instead of context.read(), I could just use anext(request_iterator) to read the messages one by one and the result will be exactly the same, right? If so, does it make sense to expose the context.read() method at all?

One month ago, Eric, Richard and I discussed this specific design. Eventually, the consistency prevails the redundancy. For new users who want to use stream API alone, this design makes them only considering the stream object itself, instead of switching the mental model between async generator and reader / writer.

> QQ: how does the python server knows if that you've chosen API1(context.write() or API2 (yield) for given method?

Yes, it does. If the function uses `yield`, it will return a generator (async generator for `async def`). We might need some reflection here, but I would say it is feasible.

> A better example

I'm trying to come up with an example that is extremely helpful when using reader / writer model, but haven't found a simple & valid one  yet. The example proposed in above comment is using the async generator API, which is not the focus point for the first example. The first example is meant to promote the usage of reader / writer API.

> I think the yield model might be difficult to use when there's not a 1 to 1 correspondence between requests and responses or when there's more logic involved

Yes, that is the main motivation of introducing reader / writer API.

> if we want to demonstrate what each of the APIs is useful for, we should use examples that really reflect that.

I agree. I always welcome suggestions 👍

Reply:
Second thought, about the `unused_request_iterator`, it can be removed if we are generating new servicer code. Currently, we don't have a detailed designed yet.